torch/_inductor/memory.py

```diff
@@ -0,0 +1,533 @@
+from __future__ import annotations
+
+import dataclasses
+import heapq
+import logging
+from typing import Callable, Dict, List, Set, Tuple, TYPE_CHECKING, Union
+
+from torch.utils._ordered_set import OrderedSet
+
+from .ir import MultiOutputLayout
+from .utils import get_dtype_size
+from .virtualized import V
+
+
+if TYPE_CHECKING:
+    from .dependencies import Dep
+    from .scheduler import BaseSchedulerNode, SchedulerBuffer
+
+
+torch_log = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class MemoryPlanningInfoForBuffer:
+    size_alloc: int = 0
+    size_free: int = 0
+    outdegree: int = 0
+    succ_nodes: OrderedSet[BaseSchedulerNode] = dataclasses.field(
+        default_factory=OrderedSet
+    )
+
+
+@dataclasses.dataclass
+class MemoryPlanningInfoForNode:
+    pred_buffers: List[Union[SchedulerBuffer, InputBuffer]] = dataclasses.field(
+        default_factory=list
+    )
+    pred_nodes: List[BaseSchedulerNode] = dataclasses.field(default_factory=list)
+    succ_nodes: List[BaseSchedulerNode] = dataclasses.field(default_factory=list)
+    indegree: int = 0
+    index: int = 0
+    memory_to_free: int = 0
+    size: int = 0
+    measure: int = 0
+
+
+@dataclasses.dataclass
+class InputBuffer:
```

Comment:
I believe there is another `InputBuffer` class in ir.py: https://github.com/pytorch/pytorch/blob/68a7246f13e8e2afd55bcf21bad6634de414b404/torch/_inductor/ir.py#L3518 and it might be slightly hard to differentiate between the two during code search. Wonder is there a way to rename this dataclass to avoid the name duplication?

Reply:
I changed the name to `FreeableInputBuffer` now.